enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+* Support allowing/denying of DNS lookups (core EgressFirewall functionality) in the new CoreDNS plugin.
+
+## Proposal
+
+This enhancement proposes to introduce a new CoreDNS plugin (``egressfirewall``) and a new Custom Resource
+(``DNSName``) to improve the integration of CoreDNS with EgressFirewall. The ``DNSName`` CR will be created
+for each unique DNS name (both regular and wildcard DNS names) used in the EgressFirewall rules. This CR will
+be used to store the DNS name along with the current IPs and the correspodning TTL and the next lookup time
+based on the TTL. The OVN-K master will be responsible for the creation of the ``DNSName`` CRs. The ``DNSName``
+CR is meant for communication between CoreDNS and OVN-K master. Thus it cannot be created or modified by an
+OpenShift cluster adminisrator.
+
+The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
+plugins. If the DNS name in the query matches any ``DNSName`` CR(s) (regular or wildcard or both), then the
+plugin will update the ``.status`` of the matching ``DNSName`` CR(s) with the DNS name along with the IPs and
+the corresponding TTL and the next lookup time based on the TTL. The OVN-K master will watch the ``DNSName``
+CRs. Whenever the IPs are updated for a ``DNSName`` CR, the OVN-K master will update the the underlying ``AdressSet``
+referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).
+
+OVN-K master will keep track of the TTL (or next lookup time) for each regular DNS name and send a DNS lookup
+query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed
+directly on the DNS name as it will not return any IP. Thus, the lookups will be performed on the DNS names
+which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.
+
+### Workflow Description
+
+The workflows for Create, Delete and Update events for EgressFirewall related to DNS names are explained in this section.
+
+#### Create/Update
+
+##### Regular DNS name
+
+* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a Namespace and adds rule(s) containing regular
+DNS name(s).
+* The OVN-K master will create corrresponding `DNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the ``openshift-dns`` Namespace and the Name of the CR will be same as the DNS
+name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.
+* The OVN-K master will then perform DNS lookup for each of the regular DNS names added to the EgressFirewall rules.
+* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookup for each of the
+regular DNS names.
+* As these DNS names have corresponding ``DNSName`` CRs, the ``egressfirewall`` plugin will update the ``.status`` of
```

Comment:
The CoreDNS plugin shouldn't be updating the DNSName CRs, the egress firewall should be.  Afaict, the DNSName CRs are relevant only to the egress firewall.